### ISSUE TITLE
Rust 1.39.0 and Async/Await

### ISSUE THREAD

**Issue by espnicholas** [clifinger/canduma#18]
Since async/await has landed with Rust 1.39.0, it would be worthwhile to take advantage of the new syntax.

**Comment by shirshak55** [comment]
@espnicholas 
hmm the problem is actix don't support async await yet?

**Comment by clifinger** [comment]
@espnicholas I will keep this issue open ... I just want to write the test first and find a better way to deal with databases. About now we should implement block to deal with Diesel sync req. I really don't know what to do/what to choose because the challenge with Canduma is to be "a framework on top of actix" with some generators (canduma console).

**Comment by espnicholas** [comment]
@shirshak55

Yeah, this is more to track against since eventually the majority of projects and crates will make a shift towards the newer syntax.

@clifinger

- Database and Diesel - Yeah that makes sense, i'll take sometime to think about it a bit.
- Actix - Agree that is the right approach. For the project overall, I know it is seen a micro-architecture for authentication but is there a desire to become more of an opinionated framework on-top of actix?

**Comment by espnicholas** [comment]
Adding https://github.com/actix/actix-net/issues/45 as a reference since this is apart of the move within Actix to async/await. I believe this work is being primarily tracked within the ticket #45 and the branch https://github.com/actix/actix-net/tree/std-future.

**Comment by shirshak55** [comment]
@espnicholas I don't think async await will be released anytime soon in actix .

**Comment by clifinger** [comment]
Hello @espnicholas and @shirshak55,

At first I read about https://github.com/rust-db/barrel to have migrations, https://docs.rs/postgres-types/0.1.0-alpha.1/postgres_types/ and https://docs.rs/tokio-postgres/0.5.0-alpha.1/tokio_postgres . What do you think about using this stack? I think it can be a good help for them since they are in alpha to have return of exp and maybe some PR from us.

@espnicholas about your question : Yes, I dream about a framework on-top of actix to easily make micro-services. I even want to tests our framework by deploying it on kubernetes (google).

**Comment by shirshak55** [comment]
@clifinger I think we should not use tokio and use async std. At least people who knows std library can easily use async std.

**Comment by clifinger** [comment]
@shirshak55 so the question is "can we adapt it" to use std?

**Comment by shirshak55** [comment]
@clifinger  I will be working for this project like 1month for full time lets see what can I bring to table. I will try to add many features that Laravel gives.

**Comment by clifinger** [comment]
@shirshak55 it's really a good news, I have also a potential good one for you ... can you add me on Skype:  ockonor (Philippines)

**Comment by shirshak55** [comment]
@clifinger hey can you let me know how can we contact each other . My email is shirshak55@gmail.com so may be we setup slack/discord for our internal communication etc?

**Comment by shirshak55** [comment]
std::future has been merged to actix-web

https://twitter.com/fafhrd91/status/1197481330277306369

**Comment by espnicholas** [comment]
@shirshak55 Yep, and looks like the examples on actix-web have been updated to async. https://github.com/actix/actix-web/blob/master/examples/basic.rs

**Comment by shirshak55** [comment]
@clifinger I am waiting for ecosystem to be abit stable so lets start once that is done.

**Comment by s-ted** [comment]
You may want to look at my attempt on this : https://github.com/s-ted/canduma-async

As I also made a few core conception changes (mostly on routing, argument passing and error management) I did not send a PR.

But all the functionalities of canduma have been migrated as async/await (i.e actix-web v2).

**Comment by shirshak55** [comment]
@s-ted looks cool :)

By the way shouldn't login logout and me should be done over graphql?

**Comment by clifinger** [comment]
@s-ted good job ! can you do a PR ? 
@shirshak55 I think it's better to separate the "security" from data exposing as he did, what do you think?

**Comment by clifinger** [comment]
BTW @s-ted 
I contacted you on linked-in for private reason, because I am from Grenoble, France :)

**Comment by shirshak55** [comment]
@clifinger I think it's better to separate the "security" from data exposing as he did? Hmm can u please elaborate I don't think there is any security issue as we can use middleware from graphql context.?

**Comment by clifinger** [comment]
@shirshak55 I don't speak about security but more about architecture, especially for micro-services.
Ex. Should we really call an API for the session in web-app? 

If you consume data from another "security" API in your GraphQl, how to send back the session?

**Comment by shirshak55** [comment]
@clifinger I am not getting it . Is there any difference between using graphql over rest api? Its same right as session token is going to be stored in redis? 

By the way are we using authorization here? I think we are only doing authentication right?

Thanks.

**Comment by clifinger** [comment]
@shirshak55 once again, there are no good or bad solutions in my opinion. Just that GraphQL does not support proxy reverse in general. Sorry for the confusing session and session cookies! Also, I do not call a simple request for authentication an API rest. For authorizations, this is a feature to come. Fell free to PR :P

**Comment by s-ted** [comment]
Hi there ! I didn't think this would bring such an enthusiasm :blush: 
I made the #19 PR, trying to explain my coding standard / decision in it as comments.

Feel free to question my decisions in the PR ;)

As for the authentication stuff in REST instead of graphql, I personaly see graphql as a good way to make "business" actions and see "business entities", but I feel authentication is more of a non-business stuff, which should be kept as a REST service, allowing non-graphql potential services (i.e. other REST services) to use it in addition to graphql.
Basically, I think that if your frontend / API consumer is graphql mature, it should be a easy for it to make the REST authentication, the opposite being not so true.

**Comment by clifinger** [comment]
Thank you so much @s-ted ... Merci beaucoup !
